File: 1415_the_k_th_lexicographical_string_of_all_happy_strings_of_length_n.cpp
class Solution { 
public:
    string getHappyString(int n, int k) {
        int u=0; string p, v;
        function<bool(string,int)> w=[&](string x,int y){
            if(x.size()==n) return ++u==k ? p=x,1 : 0;
            for(int i=0;i<3;i++) if(y!=i && w(x+char(97+i),i)) return 1;
            return 0;
        };
        return w(v,-1) ? p : "";
    }
};

File: 1415_the_k_th_lexicographical_string_of_all_happy_strings_of_length_n.java
class Solution {
    public String getHappyString(int n, int k) {
        int[] u = {0}; 
        String[] p = {""};
        return w(n, k, "", -1, u, p) ? p[0] : "";
    }

    private boolean w(int n, int k, String x, int y, int[] u, String[] p) {
        if (x.length() == n) return ++u[0] == k ? (p[0] = x) != null : false;
        for (int i = 0; i < 3; i++) 
            if (y != i && w(n, k, x + (char) (97 + i), i, u, p)) return true;
        return false;
    }
}
